cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java

```diff
@@ -601,6 +607,58 @@ public void testGetAppClasses() throws Exception {
     Assert.assertTrue(artifactStore.getApplicationClasses(Ids.namespace("ghost")).isEmpty());
   }
 
+  @Test
+  public void testExcludedPlugins() throws Exception {
+    ArtifactRange parentArtifacts = new ArtifactRange(NamespaceId.DEFAULT.getNamespace(), "parent",
+                                                      new ArtifactVersion("1.0.0"), new ArtifactVersion("2.0.0"));
+
+    // does not have any requirement
+    PluginClass includedPlugin = new PluginClass(
+      "A", "S3", "desc", "c.p2", "conf",
+      ImmutableMap.of(
+        "stream", new PluginPropertyField("stream", "description", "string", true, false)
+      )
+    );
+
+    // requires transaction
+    PluginClass excludedPlugin = new PluginClass(
+      "A", "CDAPTable", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of(Requirements.TRANSACTIONS)
+    );
+
+    // requires some other requirement than transaction
+    PluginClass includedPlugin2 = new PluginClass(
+      "A", "HbaseTable", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of("noTransactionNeeded")
+    );
+
+    // requires transaction with other requirement
+    PluginClass excludedPlugin2 = new PluginClass(
+      "A", "HbaseTable", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of("noTransactionNeeded")
+    );
+
+    Id.Artifact artifactId = Id.Artifact.from(Id.Namespace.DEFAULT, "ArtifactWithTransactionalPlugins", "1.0.0");
+    ArtifactMeta artifactMeta = new ArtifactMeta(
+      ArtifactClasses.builder().addPlugins(excludedPlugin, excludedPlugin2, includedPlugin, includedPlugin2).build(),
+      ImmutableSet.of(parentArtifacts));
+    writeArtifact(artifactId, artifactMeta, "no-content");
+    ArtifactDescriptor artifactInfo = artifactStore.getArtifact(artifactId).getDescriptor();
+
+    // plugins which have transaction as requirement should be excluded
+    Map<ArtifactDescriptor, Set<PluginClass>> actual = artifactStore.getPluginClasses(NamespaceId.DEFAULT, artifactId);
+    Assert.assertEquals(ImmutableMap.of(artifactInfo, ImmutableSet.of(includedPlugin, includedPlugin2)), actual);
+  }
```

Comment:
Should also test the other store methods. For example if you list artifacts, the ArtifactDetail should not contain any information about the plugin that should be filtered. Same for getArtifact().

Reply:
Yes, working on them right now. Will push a new commit for more test and separate one to address comments.

Reply:
Done.